CosNativeFileSystemStore.java

```diff
@@ -0,0 +1,768 @@
+        + VersionInfo.getVersion());
+
+    this.maxRetryTimes = conf.getInt(CosNConfigKeys.COSN_MAX_RETRIES_KEY,
+        CosNConfigKeys.DEFAULT_MAX_RETRIES);
+
+    config.setMaxConnectionsCount(
+        conf.getInt(CosNConfigKeys.MAX_CONNECTION_NUM,
+            CosNConfigKeys.DEFAULT_MAX_CONNECTION_NUM));
+
+    this.cosClient = new COSClient(cosCred, config);
+  }
+
+  /**
+   * Initialize the CosNativeFileSystemStore object, including
+   * its COS client and default COS bucket.
+   *
+   * @param uri  The URI of the COS bucket accessed by default.
+   * @param conf Hadoop configuration with COS configuration options.
+   * @throws IOException Initialize the COS client failed.
+   */
+  @Override
+  public void initialize(URI uri, Configuration conf) throws IOException {
+    try {
+      initCOSClient(conf);
+      this.bucketName = uri.getHost();
+    } catch (Exception e) {
+      handleException(e, "");
+    }
+  }
+
+  /**
+   * Store a file into COS from the specified input stream, which would be
+   * retried until the success or maximum number.
+   *
+   * @param key         COS object key.
+   * @param inputStream Input stream to be uploaded into COS.
+   * @param md5Hash     MD5 value of the content to be uploaded.
+   * @param length      Length of uploaded content.
+   * @throws IOException Upload the file failed.
+   */
+  private void storeFileWithRetry(String key, InputStream inputStream,
+      byte[] md5Hash, long length) throws IOException {
+    try {
+      ObjectMetadata objectMetadata = new ObjectMetadata();
+      objectMetadata.setContentMD5(Base64.encodeAsString(md5Hash));
+      objectMetadata.setContentLength(length);
+      PutObjectRequest putObjectRequest =
+          new PutObjectRequest(bucketName, key, inputStream, objectMetadata);
+
+      PutObjectResult putObjectResult =
+          (PutObjectResult) callCOSClientWithRetry(putObjectRequest);
+      LOG.debug("Store file successfully. COS key: [{}], ETag: [{}], "
+          + "MD5: [{}].", key, putObjectResult.getETag(), new String(md5Hash));
+    } catch (Exception e) {
+      String errMsg = String.format("Store file failed. COS key: [%s], "
+          + "exception: [%s]", key, e.toString());
+      LOG.error(errMsg);
+      handleException(new Exception(errMsg), key);
+    }
+  }
+
+  /**
+   * Store a local file into COS.
+   *
+   * @param key     COS object key.
+   * @param file    The local file to be uploaded.
+   * @param md5Hash The MD5 value of the file to be uploaded.
+   * @throws IOException Upload the file failed.
+   */
+  @Override
+  public void storeFile(String key, File file, byte[] md5Hash)
+      throws IOException {
+    LOG.info("Store file from local path: [{}]. file length: [{}] COS key: " +
+            "[{}] MD5: [{}].", file.getCanonicalPath(), file.length(), key,
+        new String(md5Hash));
+    storeFileWithRetry(key, new BufferedInputStream(new FileInputStream(file)),
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource created by the following constructor call: *BufferedInputStream*. Currently, there are execution paths that do not contain closure statements, e.g., when exception is thrown by ObjectMetadata constructor (in method storeFileWithRetry). Either a) close the object returned by *BufferedInputStream()* in a try-finally block or b) close the resource by declaring the object returned by *BufferedInputStream()* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).